drivers/SmartThings/jbl/src/lunchbox/rest.lua

```diff
@@ -0,0 +1,403 @@
+local socket = require "cosock.socket"
+local ssl = require "cosock.ssl"
+local log = require "log"
+
+local utils = require "lunchbox.util"
+local Request = require "luncheon.request"
+local Response = require "luncheon.response"
+
+local RestCallStates = {
+  SEND = "Send",
+  RECEIVE = "Receive",
+  RETRY = "Retry",
+  RECONNECT = "Reconnect",
+  COMPLETE = "Complete",
+}
+
+-- build a exponential backoff time value generator
+--
+-- max: the maximum wait interval (not including `rand factor`)
+-- inc: the rate at which to exponentially back off
+-- rand: a randomization range of (-rand, rand) to be added to each interval
+local function backoff_builder(max, inc, rand)
+  local count = 0
+  inc = inc or 1
+  return function()
+    local randval = 0
+    if rand then
+      randval = math.random() * rand * 2 - rand
+    end
+
+    local base = inc * (2 ^ count - 1)
+    count = count + 1
+
+    -- ensure base backoff (not including random factor) is less than max
+    if max then base = math.min(base, max) end
+
+    -- ensure total backoff is >= 0
+    return math.max(base + randval, 0)
+  end
+end
+
+local function connect(client)
+  local default_port = 80
+  local use_ssl = false
+
+  if client.base_url.scheme == "https" then
+    default_port = 443
+    use_ssl = true
+  end
+
+  local port = client.base_url.port or default_port
+
+  local sock, err = client.socket_builder(client.base_url.host, port, use_ssl)
+
+  if sock == nil then
+    client.socket = nil
+    return false, err
+  end
+
+  client.socket = sock
+  return true
+end
+
+local function reconnect(client)
+  client.socket:close()
+  client.socket = nil
+  return connect(client)
+end
+
+local function send_request(client, request)
+  local payload = request:serialize()
+
+  local bytes, err, idx = nil, nil, 0
+  if client.socket == nil then
+    err = "socket nil"
+    return bytes, err, idx
+  end
+
+  repeat bytes, err, idx = client.socket:send(payload, idx + 1, #payload) until (bytes == #payload)
+    or (err ~= nil)
+
+  return bytes, err, idx
+end
+
+local function parse_chunked_response(original_response, sock)
+  local ChunkedTransferStates = {
+    EXPECTING_CHUNK_LENGTH = "ExpectingChunkLength",
+    EXPECTING_BODY_CHUNK = "ExpectingBodyChunk",
+  }
+
+  local full_response = Response.new(original_response.status, nil)
+
+  for header in original_response.headers:iter() do full_response.headers:append_chunk(header) end
+
+  local original_body, err = original_response:get_body()
+  if not original_body or err ~= nil then
```

Comment:
A bug that was fixed in our other versions of Lunchbox was to change this from checking `original_body`'s nil status, to it's type being explicitly a `string`. That might be useful here as well.

So instead of `not original_body` you would do `type(original_body) ~= "string"`, which will also catch `nil`.

Reply:
>   if type(original_body) ~= "string" or err ~= nil then
    return original_body, (err or "unexpected nil in error position")
  end